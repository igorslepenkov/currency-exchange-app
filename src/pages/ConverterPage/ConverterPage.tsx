import { Paper, TextField, Typography } from "@mui/material";
import { Box, SxProps, Theme } from "@mui/system";
import { ChangeEvent, useEffect, useState } from "react";
import { useDebounce } from "../../hooks/useDebounce";
import { exchangeRatesService } from "../../services";
import { pageStyles } from "../../ui";
import { parseRequestString } from "../../utils";

export const ConverterPage = () => {
  const paperStyles: SxProps<Theme> = (theme) => ({
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center",
    width: "300px",
    padding: "20px",
    borderRadius: "20px",
    [theme.breakpoints.up("sm")]: {
      width: "500px",
    },
    [theme.breakpoints.up("md")]: {
      width: "700px",
    },
    [theme.breakpoints.up("lg")]: {
      width: "800px",
    },
  });

  const [input, setInput] = useState<string>("");
  const [convertResult, setConvertResult] = useState<number | null>(null);
  const [convertationError, setConvertationError] = useState<string>("");

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setInput(event.currentTarget.value);
  };

  const debouncedInputValue = useDebounce(input, 1000);

  const convertCurrencies = async (debouncedValue: string) => {
    const parsedRequest = parseRequestString(debouncedValue);
    if (parsedRequest) {
      setConvertResult(await exchangeRatesService.convert(parsedRequest));
      return;
    }

    setConvertationError(
      "We couldn't parse your request, please check if its format is right"
    );
  };

  useEffect(() => {
    if (debouncedInputValue) {
      convertCurrencies(debouncedInputValue);
    }
  }, [debouncedInputValue]);

  return (
    <Box sx={pageStyles}>
      <Paper component="form" sx={paperStyles}>
        <Typography variant="h4" mb="20px">
          Converter:
        </Typography>
        <Typography variant="body1" color="error.light" mb="20px">
          Enter your request string in format "&#60;amount&#62; &#60;base
          currency code&#62; in &#60;target currency code&#62;"
        </Typography>
        <TextField
          type="text"
          label="Your request"
          fullWidth
          onChange={handleInputChange}
          helperText={convertationError}
          FormHelperTextProps={{
            error: !!convertationError,
          }}
        />
        {convertResult && (
          <Typography sx={{ width: "100%" }} variant="h6" mt="20px" ml="10px">
            Result : {convertResult}
          </Typography>
        )}
      </Paper>
    </Box>
  );
};
