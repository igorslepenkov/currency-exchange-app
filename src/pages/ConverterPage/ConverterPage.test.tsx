import { render } from "@testing-library/react";
import { ConverterPage } from "./ConverterPage";

describe(ConverterPage, () => {
  it("Renders ConverterPage correctly", () => {
    const tree = render(<ConverterPage />);

    expect(tree).toMatchSnapshot();
  });
});
