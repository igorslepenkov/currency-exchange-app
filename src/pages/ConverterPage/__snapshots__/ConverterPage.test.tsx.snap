// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConverterPage Renders ConverterPage correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-lexdqo"
      >
        <form
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-uzqq9o-MuiPaper-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-t0sbn-MuiTypography-root"
          >
            Converter:
          </h4>
          <p
            class="MuiTypography-root MuiTypography-body1 css-x22mp-MuiTypography-root"
          >
            Enter your request string in format "&lt;amount&gt; &lt;base currency code&gt; in &lt;target currency code&gt;"
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-x22mp-MuiTypography-root"
          >
            Example: Please show me how much is 20 eur in usd
          </p>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r0:"
              id=":r0:-label"
            >
              Your request
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Your request
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-lexdqo"
    >
      <form
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-uzqq9o-MuiPaper-root"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-t0sbn-MuiTypography-root"
        >
          Converter:
        </h4>
        <p
          class="MuiTypography-root MuiTypography-body1 css-x22mp-MuiTypography-root"
        >
          Enter your request string in format "&lt;amount&gt; &lt;base currency code&gt; in &lt;target currency code&gt;"
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-x22mp-MuiTypography-root"
        >
          Example: Please show me how much is 20 eur in usd
        </p>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Your request
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Your request
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
